dwave/optimization/src/nodes/mathematical.cpp

```diff
@@ -663,6 +663,279 @@ template class NaryOpNode<functional::min<double>>;
+        : init(), array_ptr_(array_ptr), axis_(axis) {
+    if (array_ptr_->dynamic()) {
+        throw std::invalid_argument(
+                "cannot do a reduction on a dynamic array with an operation that has no identity "
+                "without supplying an initial value");
+    } else if (array_ptr_->size() < 1) {
+        throw std::invalid_argument(
+                "cannot do a reduction on an empty array with an operation that has no identity "
+                "without supplying an initial value");
+    }
+
+    /// TODO: support negative axis
+    if (axis_ < 0 || axis_ >= array_ptr_->ndim()) {
+        throw std::invalid_argument("Axis should be an integer between 0 and ndim - 1");
+    }
+
+    add_predecessor(array_ptr);
+}
+
+template <class BinaryOp>
+void PartialReduceNode<BinaryOp>::commit(State& state) const {
+    data_ptr<PartialReduceNodeData<op>>(state)->commit();
+}
+
+template <class BinaryOp>
+double const* PartialReduceNode<BinaryOp>::buff(const State& state) const {
+    return data_ptr<PartialReduceNodeData<op>>(state)->buff();
+}
+
+template <class BinaryOp>
+std::span<const Update> PartialReduceNode<BinaryOp>::diff(const State& state) const {
+    return data_ptr<PartialReduceNodeData<op>>(state)->diff();
+}
+
+template <class BinaryOp>
+void PartialReduceNode<BinaryOp>::initialize_state(State& state) const {
+    int index = topological_index();
+    assert(index >= 0 && "must be topologically sorted");
+    assert(static_cast<int>(state.size()) > index && "unexpected state length");
+    assert(state[index] == nullptr && "already initialized state");
+
+    std::vector<double> values(size(state));
+
+    for (ssize_t i = 0, stop = size(state); i < stop; ++i) {
+        assert(i < static_cast<ssize_t>(values.size()));
+        values[i] = reduce(state, i);
+    }
+
+    state[index] = std::make_unique<PartialReduceNodeData<op>>(std::move(values));
+}
+
+template <class BinaryOp>
+bool PartialReduceNode<BinaryOp>::integral() const {
+    using result_type = typename std::invoke_result<BinaryOp, double&, double&>::type;
+
+    if constexpr (std::is_integral<result_type>::value) {
+        return true;
+    }
+
+    if constexpr (std::is_same<BinaryOp, std::plus<double>>::value) {
+        // the actual value of init doesn't matter, all of the above have no default
+        // or an integer default init
+        return array_ptr_->integral() && is_integer(init.value_or(0));
+    }
+
+    assert(false && "not implemeted yet");
+    unreachable();
+}
+
+template <class BinaryOp>
+double PartialReduceNode<BinaryOp>::max() const {
+    using result_type = typename std::invoke_result<BinaryOp, double&, double&>::type;
+
+    if constexpr (std::is_same<result_type, bool>::value) {
+        return true;
+    }
+    // there are other cases we could/should handle here.
+    return Array::max();
+}
+
+template <class BinaryOp>
+double PartialReduceNode<BinaryOp>::min() const {
```

Comment:
This also needs an update along the lines of https://github.com/dwavesystems/dwave-optimization/pull/115